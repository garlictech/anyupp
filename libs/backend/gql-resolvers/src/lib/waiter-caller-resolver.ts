/* eskint-disable @typescript-eslint/no-explicit-any */
import { oeTryCatch } from '@bgap/shared/utils';
import { MutationCallWaiterArgs, CallWaiterInput, Unit } from '@bgap/domain';
import { CrudSdk } from '@bgap/crud-gql/api';
import * as OE from 'fp-ts-rxjs/lib/ObservableEither';
import { pipe } from 'fp-ts/lib/function';
import { throwError } from 'rxjs';
import { mapTo } from 'rxjs/operators';
import * as R from 'ramda';
import { AxiosStatic } from 'axios';
import { callWaiter } from '@bgap/rkeeper-api';

export interface WaiterCallerResolverDeps {
  sdk: CrudSdk;
  currentTimeISOString: () => string;
  axiosInstance: AxiosStatic;
  uuidGenerator: () => string;
}

export const getUnit =
  (deps: WaiterCallerResolverDeps) => (args: CallWaiterInput) =>
    pipe(
      deps.sdk.GetUnit({ id: args.unitId }),
      oeTryCatch,
      OE.chain(
        OE.fromPredicate(
          unit => !!unit,
          () => 'unit_not_found',
        ),
      ),
      OE.map(unit => unit as Unit),
    );

export const waiterCallerResolver =
  (deps: WaiterCallerResolverDeps) =>
  (args: MutationCallWaiterArgs): ReturnType<CrudSdk['CallWaiter']> =>
    pipe(
      args.input,
      OE.fromPredicate(
        x => !R.isEmpty(x?.unitId),
        () => 'unit_id_required',
      ),
      OE.map(x => x as CallWaiterInput),
      OE.chain(getUnit(deps)),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.canCallWaiter,
          () => 'unit_cannot_call_waiter',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.pos?.rkeeper?.endpointUri,
          () => 'not_supported_pos',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.pos?.rkeeper?.waiterOrderId,
          () => 'rkeeper_waiter_caller_not_configured',
        ),
      ),
      OE.chain(
        OE.fromPredicate(
          unit => !!unit.externalId,
          () => 'rkeeper_external_id_not_configured',
        ),
      ),
      OE.chain(unit => callWaiter(deps)(unit, args.input as CallWaiterInput)),
      OE.getOrElse<string, any>(throwError),
      mapTo(true),
    );
