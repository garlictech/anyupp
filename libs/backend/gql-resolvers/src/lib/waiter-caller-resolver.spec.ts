/* eslint-disable @typescript-eslint/no-explicit-any */
import { Observable, of, throwError } from 'rxjs';
import { waiterCallerResolver } from './waiter-caller-resolver';

describe('waiterCallerResolver unit test', () => {
  const getMockDeps = (getUnitReturn: Observable<any>) => ({
    sdk: {
      GetUnit: jest.fn().mockReturnValue(getUnitReturn),
    },
    currentTimeISOString: () => 'CURRENT TIME',
    uuidGenerator: () => 'GENERATED UUID',
    axiosInstance: {
      request: jest
        .fn()
        .mockReturnValue(Promise.resolve('RKEEPER SUCCESS RESPONSE')),
    },
  });

  test('Missing unit ID', done => {
    waiterCallerResolver({} as any)({ input: { unitId: '' } } as any).subscribe(
      {
        next: () => {
          throw 'It must fail...';
        },
        error: err => {
          expect(err).toMatchSnapshot();
          done();
        },
      },
    );
  });

  test('GetUnit error', done => {
    const deps = getMockDeps(throwError('UNIT ERROR'));
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Unit not found', done => {
    const deps = getMockDeps(of(undefined));
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Unit cannot call waiter', done => {
    const deps = getMockDeps(of({ canCallWaiter: false }));
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Not supported POS', done => {
    const deps = getMockDeps(of({ canCallWaiter: true, pos: {} }));
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Missing waiter caller config ', done => {
    const deps = getMockDeps(
      of({
        canCallWaiter: true,
        pos: {
          rkeeper: {
            endpointUri: 'ENDKEEPERURI',
          },
        },
      }),
    );
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Missing rkeeper external ID', done => {
    const deps = getMockDeps(
      of({
        canCallWaiter: true,
        pos: {
          rkeeper: {
            endpointUri: 'ENDKEEPERURI',
            waiterOrderId: 'WAITER ORDER ID',
          },
        },
      }),
    );
    waiterCallerResolver(deps as any)({
      input: { unitId: 'UNITID' } as any,
    }).subscribe({
      next: () => {
        throw 'It must fail...';
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });

  test('Send successfull waiter call', done => {
    const deps = getMockDeps(
      of({
        externalId: 'EXTERNALID',
        canCallWaiter: true,
        pos: {
          rkeeper: {
            endpointUri: 'ENDKEEPERURI',
            waiterOrderId: 'WAITER ORDER ID',
          },
        },
      }),
    );
    waiterCallerResolver(deps as any)({
      input: {
        unitId: 'UNITID',
        place: { table: '1', seat: '2' },
        guestLabel: 'GUEST LABEL',
      },
    }).subscribe({
      next: () => {
        expect(deps.axiosInstance.request.mock.calls).toMatchSnapshot(
          'AXIOS REQUEST CALLS',
        );
        done();
      },
      error: err => {
        expect(`Should not fail: ${err}`).toBeFalsy();
        done();
      },
    });
  });

  test('Send successfull waiter call', done => {
    const deps = getMockDeps(
      of({
        externalId: 'EXTERNALID',
        canCallWaiter: true,
        pos: {
          rkeeper: {
            endpointUri: 'ENDKEEPERURI',
            waiterOrderId: 'WAITER ORDER ID',
          },
        },
      }),
    );
    waiterCallerResolver(deps as any)({
      input: {
        unitId: 'UNITID',
        place: { table: '1', seat: '2' },
        guestLabel: 'GUEST LABEL',
      },
    }).subscribe({
      next: () => {
        expect(deps.axiosInstance.request.mock.calls).toMatchSnapshot(
          'AXIOS REQUEST CALLS',
        );
        done();
      },
      error: err => {
        expect(`Should not fail: ${err}`).toBeFalsy();
        done();
      },
    });
  });

  test('Rkeeper server error', done => {
    const deps = {
      ...getMockDeps(
        of({
          externalId: 'EXTERNALID',
          canCallWaiter: true,
          pos: {
            rkeeper: {
              endpointUri: 'ENDKEEPERURI',
              waiterOrderId: 'WAITER ORDER ID',
            },
          },
        }),
      ),

      axiosInstance: {
        request: jest.fn().mockImplementation(() => {
          throw 'SERVER ERROR';
        }),
      },
    };

    waiterCallerResolver(deps as any)({
      input: {
        unitId: 'UNITID',
        place: { table: '1', seat: '2' },
        guestLabel: 'GUEST LABEL',
      },
    }).subscribe({
      next: () => {
        expect('Must throw error').toBeFalsy();
        done();
      },
      error: err => {
        expect(err).toMatchSnapshot();
        done();
      },
    });
  });
});
