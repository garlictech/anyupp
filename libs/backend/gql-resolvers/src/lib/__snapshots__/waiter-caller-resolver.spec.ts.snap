// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`waiterCallerResolver unit test GetUnit error 1`] = `"UNIT ERROR"`;

exports[`waiterCallerResolver unit test Missing rkeeper external ID 1`] = `"rkeeper_external_id_not_configured"`;

exports[`waiterCallerResolver unit test Missing unit ID 1`] = `"unit_id_required"`;

exports[`waiterCallerResolver unit test Missing waiter caller config  1`] = `"rkeeper_waiter_caller_not_configured"`;

exports[`waiterCallerResolver unit test Not supported POS 1`] = `"not_supported_pos"`;

exports[`waiterCallerResolver unit test Rkeeper server error 1`] = `"SERVER ERROR"`;

exports[`waiterCallerResolver unit test Send successfull waiter call: AXIOS REQUEST CALLS 1`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "password": "",
        "username": "",
      },
      "data": Object {
        "client": Object {
          "email": undefined,
          "ln": "Biborka",
          "phone": undefined,
        },
        "delivery_time": "CURRENT TIME",
        "guest_label": 2,
        "guest_name": "GUEST LABEL",
        "objectid": "EXTERNALID",
        "order": Object {
          "id": "WAITER ORDER ID",
          "qnt": 1,
          "type": "d",
        },
        "order_number": 1,
        "order_type": 1,
        "pay_online_type": 0,
        "pay_type": 0,
        "remoteId": "EXTERNALID",
        "remoteOrderId": "GENERATED UUID",
      },
      "method": "post",
      "url": "ENDKEEPERURI/postorder/EXTERNALID",
    },
  ],
]
`;

exports[`waiterCallerResolver unit test Send successfull waiter call: AXIOS REQUEST CALLS 2`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "password": "",
        "username": "",
      },
      "data": Object {
        "client": Object {
          "email": undefined,
          "ln": "Biborka",
          "phone": undefined,
        },
        "delivery_time": "CURRENT TIME",
        "guest_label": 2,
        "guest_name": "GUEST LABEL",
        "objectid": "EXTERNALID",
        "order": Object {
          "id": "WAITER ORDER ID",
          "qnt": 1,
          "type": "d",
        },
        "order_number": 1,
        "order_type": 1,
        "pay_online_type": 0,
        "pay_type": 0,
        "remoteId": "EXTERNALID",
        "remoteOrderId": "GENERATED UUID",
      },
      "method": "post",
      "url": "ENDKEEPERURI/postorder/EXTERNALID",
    },
  ],
]
`;

exports[`waiterCallerResolver unit test Unit cannot call waiter 1`] = `"unit_cannot_call_waiter"`;

exports[`waiterCallerResolver unit test Unit not found 1`] = `"unit_not_found"`;
